### PR TITLE
Auto register/unregister hooks in gen_mod

### PR DESCRIPTION
**Pull request by kamilwaz**
To keep it small, I edited only mods that were very easy to transform.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/164150) / small_tests / 042d99c9cb7c86f8faeaa77f69e056c49d060efb
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3954/164150/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3954/164150/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-01-30_12.42.23/lib.mongooseim.logs/run.2023-01-30_12.42.24/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/164441) / small_tests / 693ac88e500d92bdbfe186b3e30400c6bc76a8c5
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3954/164441/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3954/164441/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-01-31_10.38.52/lib.mongooseim.logs/run.2023-01-31_10.38.53/suite.log.html)

**Review by chrzaszcz — COMMENTED** [review]
It looks good to me in general, the main benefit is that you don't have to remember to register and unregister the hooks anymore. Apart from minor comments, the only issue I found was that sometimes the hooks were registered before some startup operations on purpose to avoid having missing handlers if the module starts handling traffic. On the other hand, such race conditions are very unlikely and extremely short-lived. Actually I found it only in one module (see comments), while for some modules this PR actually fixes the startup order.

While we are still deciding on the callback format, one could pass `Opts` as the second argument, allowing modules like `mod_global_distrib` to set up correct hooks. It is possible to access module opts when it is being stopped, and pass them there. However, I like the simplicity of the one-argument callback, so I think it's good as it is in the current PR, and we could always add the two-argument one as another option.

File: src/muc_light/mod_muc_light.erl [hunk 1]
```diff
@@ -174,7 +174,6 @@ start(HostType, Opts) ->
     Codec = host_type_to_codec(HostType),
     mod_muc_light_db_backend:start(HostType, Opts),
     mod_muc_light_codec_backend:start(HostType, #{backend => Codec}),
-    gen_hook:add_handlers(hooks(HostType)),
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/muc_light/mod_muc_light.erl:177]: Actually `mongoose_listener` starts after `mongoose_modules`, which makes this race condition impossible to occur in a production environment, and in the tests we start modules before sending any traffic. The code in this PR is good as it is.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/165298) / small_tests / 7ad2a8c086a045f0e08615581815970220e27dfa
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3954/165298/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3954/165298/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-07_12.34.59/lib.mongooseim.logs/run.2023-02-07_12.35.00/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/165610) / small_tests / a55aa37920d4d15b3b9e34ece4771696cf4ddc47
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3954/165610/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3954/165610/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-13_14.04.09/lib.mongooseim.logs/run.2023-02-13_14.04.09/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/165746) / small_tests / 2da25195cb4f750e5226734f4395e731e7ba7206
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3954/165746/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3954/165746/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-14_12.55.19/lib.mongooseim.logs/run.2023-02-14_12.55.19/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/166254) / small_tests / 096229066e7c476608ad9d2ca716d62f47da7b6d
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3954/166254/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3954/166254/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-17_11.41.59/lib.mongooseim.logs/run.2023-02-17_11.41.59/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/166278) / small_tests / 906bbfb1d413a53e31215fb2333c5d98d170fe5a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3954/166278/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3954/166278/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-17_11.56.41/lib.mongooseim.logs/run.2023-02-17_11.56.42/suite.log.html)

**Review by arcusfelis — APPROVED** [review]
looks ok

File: src/mod_csi.erl [hunk 2]
```diff
@@ -34,8 +34,8 @@ start(HostType, _Opts) ->
     gen_hook:add_handlers(hooks(HostType)).
```
Comment (chrzaszcz) [inline src/mod_csi.erl:34]: Double registration?

File: test/keystore_SUITE.erl [hunk 3]
```diff
@@ -22,46 +22,43 @@ init_per_suite(C) ->
     C.
 
 end_per_suite(C) ->
+    [mongoose_config:unset_opt(Key) || {Key, _Value} <- opts()],
+    [mongoose_config:unset_opt({modules, Host})|| Host <- hosts()],
```
Comment (chrzaszcz) [inline test/keystore_SUITE.erl:26]:
```suggestion
    [mongoose_config:unset_opt({modules, Host}) || Host <- hosts()],
```

**Review by chrzaszcz — COMMENTED** [review]
I added some comments. There are unresolved comments from the previous code review as well. Please explain the "fix unit tests" commit including the motivation for using the process dictionary.

**Comment by kamilwaz** [comment]
> Please explain the "fix unit tests" commit including the motivation for using the process dictionary.

I added this to let developers to track processes started via `async_helper`. `test_server_loc` is set by `ct` and should include a suite name. This may be helpful to check which tests start a process and don't shutdown it later.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/167436) / small_tests / 95d83b5b8e0ffbc510fb6a6b15a09d2c2cf9b65d
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3954/167436/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3954/167436/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-03-01_12.27.50/lib.mongooseim.logs/run.2023-03-01_12.27.51/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/167493) / small_tests / 1e82fc4d3ee33d825066a544dca2d694efa284f6
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3954/167493/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3954/167493/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-03-01_14.36.43/lib.mongooseim.logs/run.2023-03-01_14.36.43/suite.log.html)

**Review by chrzaszcz — APPROVED** [review]
Looks good 👍

**Pull request merged**